pennylane/math/matrix_manipulation.py

```diff
@@ -105,40 +106,86 @@ def expand_matrix(base_matrix, wires, wire_order=None, sparse_format="csr"):
         return base_matrix
 
     interface = qml.math.get_interface(base_matrix)  # pylint: disable=protected-access
+
     if interface == "scipy" and issparse(base_matrix):
         return _sparse_expand_matrix(base_matrix, wires, wire_order, format=sparse_format)
 
-    wire_order = qml.wires.Wires(wire_order)
-    n = len(wires)
+    wire_order = list(wire_order)
+
     shape = qml.math.shape(base_matrix)
     batch_dim = shape[0] if len(shape) == 3 else None
 
-    # operator's wire positions relative to wire ordering
-    op_wire_pos = wire_order.indices(wires)
-
-    identity = qml.math.reshape(
-        qml.math.eye(2 ** len(wire_order), like=interface), [2] * (len(wire_order) * 2)
-    )
-    # The first axis entries are range(n, 2n) for batch_dim=None and range(n+1, 2n+1) else
-    axes = (list(range(-n, 0)), op_wire_pos)
-
-    # reshape op.matrix()
-    op_matrix_interface = qml.math.convert_like(base_matrix, identity)
-    shape = [batch_dim] + [2] * (n * 2) if batch_dim else [2] * (n * 2)
-    mat_op_reshaped = qml.math.reshape(op_matrix_interface, shape)
-    mat_tensordot = qml.math.tensordot(
-        mat_op_reshaped, qml.math.cast_like(identity, mat_op_reshaped), axes
-    )
-
-    unused_idxs = [idx for idx in range(len(wire_order)) if idx not in op_wire_pos]
-    # permute matrix axes to match wire ordering
-    perm = op_wire_pos + unused_idxs
-    sources = wire_order.indices(wire_order)
-    if batch_dim:
-        perm = [p + 1 for p in perm]
-        sources = [s + 1 for s in sources]
-
-    mat = qml.math.moveaxis(mat_tensordot, sources, perm)
+    expanded_wires = []
+    mats = []  # list of tuples with (Id,) and (base_matrix,) or (mat1, mat2, ...) if batch_dim
+    op_wires_in_list = False
+    i_count = 0  # identity count
+    for wire in wire_order:
+        if wire not in wires:
+            i_count += 1
+            expanded_wires.append(wire)
+        elif not op_wires_in_list:
+            if i_count > 0:
+                mats.append(
+                    (qml.math.cast_like(qml.math.eye(2**i_count, like=interface), base_matrix),)
+                )
+            i_count = 0
+            mats.append(tuple(base_matrix) if batch_dim else (base_matrix,))
+            op_wires_in_list = True
+            expanded_wires.extend(wires)
+
+    if i_count > 0:
+        mats.append((qml.math.cast_like(qml.math.eye(2**i_count, like=interface), base_matrix),))
```

Comment:
I wonder if we could extract this logic and use it in both dense and sparse matrix methods?